File: devops/linux/syscall/src/process_management_examples.c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <sys/wait.h>

int main() {
    printf("Main process PID: %d\n", getpid());
    printf("Parent process PID: %d\n", getppid());

    pid_t pid = fork(); // Create child process
    
    if (pid < 0) {
        perror("fork failed");
        exit(1);
    }
    else if (pid == 0) {
        printf("Child process (PID: %d, Parent PID: %d)\n", getpid(), getppid());
        
        printf("Child executing 'ls -l'...\n");
        execl("/bin/ls", "ls", "-l", NULL);
        
        // If execl returns, it failed
        perror("execl failed");
        exit(1);
    } 
    else {
        printf("Parent process (PID: %d) waiting for child (PID: %d)\n", getpid(), pid);
        
        int status;
        pid_t wait_pid = wait(&status);
        
        if (WIFEXITED(status)) {
            printf("Child %d exited with status %d\n", wait_pid, WEXITSTATUS(status));
        }
    }

    return 0;
}

File: devops/linux/syscall/src/process_management_examples.py
import os
import sys

def main():
    print(f"Main process PID: {os.getpid()}")
    print(f"Parent process PID: {os.getppid()}")

    pid = os.fork() # Create child process
    
    if pid < 0:
        print("fork failed", file=sys.stderr)
        sys.exit(1)
    elif pid == 0:
        print(f"Child process (PID: {os.getpid()}, Parent PID: {os.getppid()})")
        
        print("Child executing 'ls -l'...")
        os.execl('/bin/ls', 'ls', '-l')
        
        # If execl returns, it failed
        print("execl failed", file=sys.stderr)
        sys.exit(1)
    else:
        # Parent process
        print(f"Parent process (PID: {os.getpid()}) waiting for child (PID: {pid})")
        
        pid, status = os.wait()
        if os.WIFEXITED(status):
            print(f"Child {pid} exited with status {os.WEXITSTATUS(status)}")

if __name__ == "__main__":
    main()